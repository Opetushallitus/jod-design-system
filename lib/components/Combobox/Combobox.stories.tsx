import type { Meta, StoryObj } from '@storybook/react';

import { Combobox } from './Combobox';

const meta = {
  title: 'Combobox',
  component: Combobox,
  tags: ['autodocs'],
} satisfies Meta<typeof Combobox>;

export default meta;

type Story = StoryObj<typeof meta>;
const label = 'Choose language';
const decorators: Story['decorators'] = [
  (Story) => (
    <div className="ds-min-h-[250px]">
      <Story />
    </div>
  ),
];
const design = {
  type: 'figma',
  url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-7243&t=RDNOudQv5NR9oxZQ-4',
};

export const Default: Story = {
  decorators,
  parameters: {
    design,
    docs: {
      description: {
        story: 'Combobox component for selecting a value from a list.',
      },
    },
  },
  args: {
    label,
    placeholder: label,
    options: [
      { label: 'Suomi', value: 'fi' },
      { label: 'Svenska', value: 'sv' },
      { label: 'English', value: 'en' },
    ],
  },
};

export const Disabled: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story: 'Disabled combobox component. User cannot interact with the component.',
      },
    },
  },
  args: {
    disabled: true,
    label,
    placeholder: label,
    options: [
      { label: 'Suomi', value: 'fi' },
      { label: 'Svenska', value: 'sv' },
      { label: 'English', value: 'en' },
    ],
  },
};

export const PreSelectedValue: Story = {
  decorators,
  parameters: {
    design,
    docs: {
      description: {
        story: 'Combobox component with preselected value',
      },
    },
  },
  args: {
    label,
    placeholder: label,
    selected: 'en',
    options: [
      { label: 'Suomi', value: 'fi' },
      { label: 'Svenska', value: 'sv' },
      { label: 'English', value: 'en' },
    ],
  },
};
