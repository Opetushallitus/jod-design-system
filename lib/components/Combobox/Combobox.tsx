import {
  ComboboxButton,
  ComboboxInput,
  ComboboxOption,
  ComboboxOptions,
  Combobox as HeadlessCombobox,
} from '@headlessui/react';
import React from 'react';
import { MdExpandLess, MdExpandMore } from 'react-icons/md';

export interface ComboboxOptionsData<T extends string = string> {
  value: T;
  label: string;
}

interface ComboboxProps<T extends ComboboxOptionsData, U extends string = string> {
  /** Label for the component */
  label: string;
  /** Hide label. Still available for screenreaders */
  hideLabel?: boolean;
  /** Options for component */
  options: T[];
  /** Default value to be selected initially */
  defaultValue?: U;
  /** Controlled mode */
  selected?: U;
  /** Callback on selection change */
  onChange?: (value: U) => void;
  /** Component is disabled for user interaction */
  disabled?: boolean;
  /** Placeholder text */
  placeholder: string;
}

export const Combobox = <
  U extends string = string,
  T extends ComboboxOptionsData<string> = ComboboxOptionsData<string>,
>({
  label,
  hideLabel = false,
  options,
  defaultValue,
  onChange: propOnChange,
  placeholder,
  selected,
  disabled = false,
}: ComboboxProps<T, U>) => {
  const labelId = React.useId();

  const [query, setQuery] = React.useState('');
  const filteredOptions =
    query === ''
      ? options
      : options.filter((option) => {
          return option.label.toLowerCase().includes(query.toLowerCase());
        });

  return (
    <div className="ds:flex ds:flex-col ds:relative">
      {!hideLabel && (
        <label htmlFor={labelId} className="ds:text-jod-black ds:text-form-label ds:font-arial ds:mb-4">
          {label}
        </label>
      )}
      <div className="ds:flex ds:flex-row relative">
        <HeadlessCombobox
          defaultValue={defaultValue ?? (options[0]?.value as U)}
          onChange={propOnChange}
          disabled={disabled}
          value={selected}
        >
          {({ open }) => (
            <>
              <ComboboxInput
                id={labelId}
                aria-label={hideLabel ? label : undefined}
                displayValue={(value: U) => options.find((option) => option.value === value)?.label ?? ''}
                className="ds:font-arial ds:w-full ds:rounded-l ds:border-y ds:border-l ds:border-border-gray ds:bg-white ds:p-5 ds:text-black ds:outline-hidden ds:placeholder:text-inactive-gray ds:disabled:text-inactive-gray ds:disabled:pointer-events-none"
                onChange={(event) => setQuery(event.target.value)}
                placeholder={`(${placeholder})`}
              />
              <ComboboxButton
                className="ds:select-none ds:rounded-r ds:border-y ds:border-r ds:border-border-gray ds:bg-white ds:p-5 ds:text-secondary-gray ds:disabled:text-inactive-gray"
                disabled={disabled}
              >
                {open ? <MdExpandLess size={24} /> : <MdExpandMore size={24} />}
              </ComboboxButton>
              <ComboboxOptions className="ds:bg-white ds:mt-3 ds:absolute ds:w-full ds:top-full ds:p-5 ds:m-0 ds:shadow-border ds:rounded-md ds:z-50 ds:empty:invisible">
                {filteredOptions.map((option) => (
                  <ComboboxOption
                    key={option.value}
                    className="ds:py-3 ds:text-heading-4 ds:ml-5 ds:text-black ds:cursor-pointer ds:data-focus:underline ds:data-focus:text-accent ds:hover:underline ds:hover:text-accent"
                    value={option.value}
                  >
                    {option.label}
                  </ComboboxOption>
                ))}
              </ComboboxOptions>
            </>
          )}
        </HeadlessCombobox>
      </div>
    </div>
  );
};
