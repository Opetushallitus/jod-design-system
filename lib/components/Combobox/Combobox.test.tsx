import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { Combobox } from './Combobox';

describe('Combobox', () => {
  const options = [
    { value: 'fi', label: 'Suomi' },
    { value: 'sv', label: 'Svenska' },
    { value: 'en', label: 'English' },
  ];

  const placeholder = 'placeholder text';

  describe('Snapshot testing', () => {
    it('should render with defaults', () => {
      const { container } = render(<Combobox label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should not show label', () => {
      const { container } = render(<Combobox hideLabel label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render as disabled', () => {
      const { container } = render(<Combobox disabled label="Label" options={options} placeholder={placeholder} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('defaultValue', () => {
    it('should select first by default', async () => {
      render(<Combobox label="Label" options={options} placeholder={placeholder} />);
      const input: HTMLInputElement = screen.getByRole('combobox');
      expect(input.value).toBe('Suomi');
      const user = userEvent.setup();
      await user.click(screen.getByRole('button'));
      expect(screen.getByRole('option', { name: 'Suomi' }).ariaSelected).toBe('true');
    });

    it('should select the wanted one', async () => {
      render(<Combobox label="Label" options={options} placeholder={placeholder} defaultValue="sv" />);
      const input: HTMLInputElement = screen.getByRole('combobox');
      expect(input.value).toBe('Svenska');
      const user = userEvent.setup();
      await user.click(screen.getByRole('button'));
      expect(screen.getByRole('option', { name: 'Svenska' }).ariaSelected).toBe('true');
    });

    it('should still allow user to another', async () => {
      render(<Combobox label="Label" options={options} placeholder={placeholder} defaultValue="fi" />);
      const input: HTMLInputElement = screen.getByRole('combobox');
      // Click the expand button and select English
      const user = userEvent.setup();
      await user.click(screen.getByRole('button'));
      await user.selectOptions(screen.getByRole('listbox'), 'English');

      // Assert that the value is English, click the expand button and check that English is selected
      expect(input.value).toBe('English');
      await user.click(screen.getByRole('button'));
      expect(screen.getByRole('option', { name: 'English' }).ariaSelected).toBe('true');
    });
  });

  describe('onChange', () => {
    const options = [
      { value: 'fi', label: 'Suomi' },
      { value: 'sv', label: 'Svenska' },
      { value: 'en', label: 'English' },
    ];

    it('is called on user selection', async () => {
      const mockCallback = vi.fn();
      const user = userEvent.setup();

      render(<Combobox label="Label" options={options} placeholder={placeholder} onChange={mockCallback} />);
      await user.click(screen.getByRole('button'));
      await user.selectOptions(screen.getByRole('listbox'), 'English');
      expect(mockCallback).toHaveBeenCalledTimes(1);
    });
  });

  describe('label', () => {
    it('should render with label', () => {
      render(<Combobox label="Label" options={[]} placeholder={placeholder} />);
      expect(screen.getAllByText('Label')).not.toBeNull();
    });
  });

  describe('hideLabel', () => {
    it('should add aria-label when label is hidden', () => {
      render(<Combobox label="Non-visible label" options={[]} hideLabel={true} placeholder={placeholder} />);
      expect(screen.getByRole('combobox', { name: 'Non-visible label' })).not.toBeNull();
    });
  });
});
