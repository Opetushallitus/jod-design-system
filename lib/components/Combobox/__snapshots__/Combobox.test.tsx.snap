// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Combobox > Snapshot testing > should not show label 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <div
    class="ds:flex ds:flex-row relative"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      aria-label="Label"
      class="ds:font-arial ds:w-full ds:rounded-l ds:border-y ds:border-l ds:border-border-gray ds:bg-white ds:p-5 ds:text-black ds:outline-hidden ds:placeholder:text-inactive-gray ds:disabled:text-inactive-gray ds:disabled:pointer-events-none"
      data-headlessui-state=""
      id=":r8:"
      placeholder="(placeholder text)"
      role="combobox"
      type="text"
      value="Suomi"
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="ds:select-none ds:rounded-r ds:border-y ds:border-r ds:border-border-gray ds:bg-white ds:p-5 ds:text-secondary-gray ds:disabled:text-inactive-gray"
      data-headlessui-state=""
      id="headlessui-combobox-button-:rc:"
      tabindex="-1"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Combobox > Snapshot testing > should render as disabled 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <label
    class="ds:text-jod-black ds:text-form-label ds:font-arial ds:mb-4"
    for=":rg:"
  >
    Label
  </label>
  <div
    class="ds:flex ds:flex-row relative"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="ds:font-arial ds:w-full ds:rounded-l ds:border-y ds:border-l ds:border-border-gray ds:bg-white ds:p-5 ds:text-black ds:outline-hidden ds:placeholder:text-inactive-gray ds:disabled:text-inactive-gray ds:disabled:pointer-events-none"
      data-disabled=""
      data-headlessui-state="disabled"
      disabled=""
      id=":rg:"
      placeholder="(placeholder text)"
      role="combobox"
      type="text"
      value="Suomi"
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="ds:select-none ds:rounded-r ds:border-y ds:border-r ds:border-border-gray ds:bg-white ds:p-5 ds:text-secondary-gray ds:disabled:text-inactive-gray"
      data-disabled=""
      data-headlessui-state="disabled"
      disabled=""
      id="headlessui-combobox-button-:rk:"
      tabindex="-1"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Combobox > Snapshot testing > should render with defaults 1`] = `
<div
  class="ds:flex ds:flex-col ds:relative"
>
  <label
    class="ds:text-jod-black ds:text-form-label ds:font-arial ds:mb-4"
    for=":r0:"
  >
    Label
  </label>
  <div
    class="ds:flex ds:flex-row relative"
  >
    <input
      aria-autocomplete="list"
      aria-expanded="false"
      class="ds:font-arial ds:w-full ds:rounded-l ds:border-y ds:border-l ds:border-border-gray ds:bg-white ds:p-5 ds:text-black ds:outline-hidden ds:placeholder:text-inactive-gray ds:disabled:text-inactive-gray ds:disabled:pointer-events-none"
      data-headlessui-state=""
      id=":r0:"
      placeholder="(placeholder text)"
      role="combobox"
      type="text"
      value="Suomi"
    />
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="ds:select-none ds:rounded-r ds:border-y ds:border-r ds:border-border-gray ds:bg-white ds:p-5 ds:text-secondary-gray ds:disabled:text-inactive-gray"
      data-headlessui-state=""
      id="headlessui-combobox-button-:r4:"
      tabindex="-1"
      type="button"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
        />
      </svg>
    </button>
  </div>
</div>
`;
